src/main/java/me/damonkelley/tictactoe/finder/ArtificialIntelligenceFinder.java

```diff
@@ -1,61 +1,74 @@
         return choice;
     }
 
-    private int minimax(State state, int depth, boolean maximizingPlayer) {
-        if (new GameRules(state).isOver() || depth == 0) return scoreFor(state, depth);
+    private int minimax(Game game, int depth, int alpha, int beta, boolean maximizingPlayer) {
+        if (game.isOver() || depth == 0) return scoreFor(game, depth);
 
         HashMap<Integer, Space> scores = new HashMap<>();
 
-        for (Space space : state.getBoard().availableSpaces()) {
-            State futureState = state.copy();
-            futureState.move(space, futureState.getNextMarker());
+        if (maximizingPlayer) {
+            int score = alpha;
+            for (Space space : game.getBoard().availableSpaces()) {
+                Game futureGame = game.copy();
+                futureGame.move(space, futureGame.nextTurn());
 
-            scores.put(minimax(futureState, depth - 1, !maximizingPlayer), space);
-        }
+                score = Integer.max(score, minimax(futureGame, depth - 1, alpha, beta, !maximizingPlayer));
+                scores.putIfAbsent(score, space);
 
-        int bestScore = bestScoreFor(maximizingPlayer, scores);
-        choice = scores.get(bestScore);
+                alpha = Integer.max(alpha, score);
+                if (beta <= alpha)
+                    break;
+            }
 
-        return bestScore;
-    }
+            choice = scores.get(findMaximumScore(scores));
+            return score;
+        } else {
+            int score = beta;
+            for (Space space : game.getBoard().availableSpaces()) {
+                Game futureGame = game.copy();
+                futureGame.move(space, futureGame.nextTurn());
 
-    private int scoreFor(State state, int depth) {
-        GameRules rules = new GameRules(state);
-        if (rules.determineWinner() == marker) {
-            return 10 + depth;
-        } else if (rules.isOver() && !rules.isDraw()) {
-            return -10 - depth;
+                score = Integer.min(score, minimax(futureGame, depth - 1, alpha, beta, !maximizingPlayer));
+                scores.putIfAbsent(score, space);
+
+                beta = Integer.min(score, beta);
+                if (beta <= alpha)
+                    break;
+            }
+
+            choice = scores.get(findMinimumScore(scores));
+            return score;
         }
-        return 0;
     }
 
-    private int bestScoreFor(boolean maximizingPlayer, HashMap<Integer, Space> scores) {
-        if (maximizingPlayer) {
-            return findMaximumScore(scores);
-        } else {
-            return findMinimumScore(scores);
+    private int scoreFor(Game game, int depth) {
+        if (game.determineWinner() == marker) {
+            return 10 + depth;
+        } else if (game.isOver() && !game.isDraw()) {
+            return -10 - depth;
         }
+        return depth;
     }
 
     private Integer findMinimumScore(HashMap<Integer, Space> scores) {
```

Comment:
It looks like this data structure isn't doing you any favors in terms of read performance. You might consider looking at a sorted tree data structure like [`TreeMap`](https://docs.oracle.com/javase/8/docs/api/java/util/TreeMap.html) so that you're more like O(1) than O(N) in finding the max/min score.